### ISSUE TITLE
Cannot discard unstaged changes

### ISSUE THREAD

**Issue by joonoro** [magit/magit#1960]
You used to be able to discard *unstaged* changes just by pressing `v` on a hunk to call `magit-revert-item`. However the function has since been removed and `v` is now bound to `magit-reverse` which can only discard *staged* changes. Pressing it on an unstaged hunk says:

    user-error: Cannot reverse unstaged changes

Is this a bug, or is there a new way to discard unstaged changes?

**Comment by npostavs** [comment]
`k` now discards both staged and unstaged changes.

**Comment by joonoro** [comment]
Ah, thanks for the quick answer. I guess I should have read the updated manual as I just noticed it's mentioned there.

**Comment by robohack** [comment]
The question this issue raised bothers me, and I wouldn't have even found the reference in the manual had I not searched for `magit-reverse` here in the issues list because I ran into the same problem.  I didn't even know about the new `magit-discard` before now either.

The manual doesn't seem clear to me on the distinction between "reverse" and "discard", especially with respect to un-staged changes.

The help strings for both `magit-reverse` and `magit-discard` are no better.

It also seems less than helpful to have a command that used to just "do the right thing" (DWIM) stop working and complain, especially when the complaint is rather meaningless and content free and without any helpful suggestions as to alternatives.

To an uninformed user (me) it would seem both 'k' and 'v' could and should do the same thing, at least for un-staged changes, and probably for staged changes, and maybe even for committed changes too.  I just want to undo a change (get rid of it, initially just from the working directory) by putting the cursor at a hunk in a diff (or on a file for all hunks in the file, or ideally someday on one of a set of marked hunks possibly across several files), and I don't really care if it's un-staged, staged, or committed -- the same key should work in all cases, and I don't care if the operation is called 'revert', 'reverse', 'undo', or 'discard' -- they all would seem to ultimately mean the same thing to me.  (It would seem obvious to me that for undoing committed changes the result should just be left as a new un-staged change.)

As a semi-informed user (almost-me), the only distinction I can see is when a change is staged vs. un-staged.  In the 'staged' case there could be a question as to whether the change should be entirely removed, or just un-staged, but on the other hand I think 'stage' ('s') and 'un-stage' ('u') are relatively easy to understand and keep separate, and so I think getting entirely rid of the change from the working directory is obviously different enough from un-staging a change as to not cause confusion.  Getting rid of the change means removing it from the working directory, which should, I think, also clearly enough mean that it must simultaneously be un-staged (don't ever allow a commit of something not in the working directory); while of course un-staging just means taking it away from the set of change hunks to be committed and not changing the working directory.

BTW, I have no clue as to whether these "reverse" and "discard" ideas come from somewhere in Git or not -- I'm only seeing them from the Magit user's perspective.  Any time I have to stop and think about which key I have hit to achieve the same ultimate goal then I'm going to get confused, and I'm going to get frustrated when I have to read the doc string and/or the manual to figure out the differences.  In theory even 'V v' should work the same on un-staged or staged changes as it would for committed changes.

**Comment by tarsius** [comment]
First things first: I agree that the documentation on these essential features is a bit sparse and that it should be extended. Never-the-less even in its current form the manual contains all that has to be known - its just very concise.

> I don't care if the operation is called 'revert', 'reverse', 'undo', or 'discard' -- they all would seem to ultimately mean the same thing to me.

This seems to be the root cause of your difficulties to understand these concepts. I very carefully chose the terms `discard` and `reverse`. By choosing not to reason about, or to even ignore the terminology you deprive yourself of the opportunity to deduce meaning from it. That's very unfortunate, because the documentation was written on the premise that the user would read it while keeping the general meaning of these words in mind.

> I have no clue as to whether these "reverse" and "discard" ideas come from somewhere in Git or not

Both of these concepts do not exist in a *pure* form in Git. Git doesn't use the term "discard" at all I think, it's my "invention", and I think I did a very good job picking the appropriate term for it. "reverse" comes from `git apply --reverse`, it's not quite as good a term because of the ambiguity between "reverse" and "revert". But when contrasting it with "discard" it gets the job done, and using Git terminology were it already exists also has its advantages.

From the "Oxford Dictionaries":

> [reverse](http://www.oxforddictionaries.com/definition/english/reverse)
> 2. Make (something) the opposite of what it was: the damage done to the ozone layer may be reversed
> 3. Turn (something) the other way round or up or inside out

> [discard](http://www.oxforddictionaries.com/definition/english/discard)
> 1 Get rid of (someone or something) as no longer useful or desirable

Even if mankind figures out how to reverse the effect it has had on the ecosystem, that doesn't change the historic fact that we did do the damage. On the other hand if you have guests coming over for dinner and put some flowers on the table but then remember that one of them has an allergy and discard the flowers in the trash, then it is as if the flowers never were on the table.

Similarly when you "reverse" a change in Magit, this is something very different from "discarding" it. When you "reverse" a change, then you take something that happened in the past, and then do the opposite in the presence. After you have done that, the original event still exists, you have only done another change, which reverses its **effect**. "Reversing" means to take a change from one place, the source (ie the past aka a commit (or the almost-commit called index)), and apply it in reverse to another place, the destination (ie the presence aka the working copy). The original change still exists in the source.

Contrast that with "discarding" a change. Here there is no source and destination, there is only one place (the working copy and/or index). "Discarding" a change means to go to the place where the change exists, take it, and throw it away without saving it, or something deriving from it, anywhere else. After you have "discarded" a change then it is gone without leaving any traces behind.

> It also seems less than helpful to have a command that used to just "do the right thing" (DWIM) stop working and complain, 

What I have said above hopefully explains that "discarding" a change is inherently more dangerous than "reversing" it. (A reverse action can always be undone by discarding it, while a discard action is final). So the old behavior of having "reverse" fall back to "discard" in situations where "reversing" is not possible was not a feature, it was a grave bug which could lead to data loss.

Not only could it lead to data loss, I think that this old inconsistent behavior also contributes to your current misunderstanding, and possibly to that of others too.

> (don't ever allow a commit of something not in the working directory);

No, allowing to do just that is a very useful, though advanced, feature. See https://github.com/magit/magit/wiki/History-Manipulation for a use case.
